### ISSUE TITLE
Do not flag self, static nor this in an anonymous function declared in class scope

### ISSUE THREAD

**Issue by david-binda** [Automattic/VIP-Coding-Standards#82]
Related code: https://github.com/Automattic/VIP-Coding-Standards/blob/master/WordPressVIPMinimum/Sniffs/Variables/VariableAnalysisSniff.php#L853

> As of PHP 5.4.0, when declared in the context of a class, the current class is automatically bound to it, making $this available inside of the function's scope. If this automatic binding of the current class is not wanted, then static anonymous functions may be used instead.
via http://php.net/manual/en/functions.anonymous.php#example-167

Example code producing false-positives:

```
<?php

abstract class My_Class {

	protected static $_type = '';

	public static function method() {
		add_action(
			'admin_notices',
			function() {
				self::admin_notice( 'Please pick another $_type for ' . get_called_class() . '. <em><strong>' . static::$_type . '</strong></em> is already registered.' );
			}
		);
	}
}

```

**Comment by GaryJones** [comment]
@david-binda This is something covered in PHPCompatibility: https://github.com/PHPCompatibility/PHPCompatibility/pull/669 - can we close this Issue out?

**Comment by GaryJones** [comment]
I'm going to presumptively close this out. Can be re-opened if necessary.

**Comment by rebeccahum** [comment]
Re-opening because this is flagging false positives for `self::` being used inside a class scope via `WordPressVIPMinimum.Variables.VariableAnalysis.StaticInsideClosure`.

**Comment by GaryJones** [comment]
> flagging false positives

Can you give some example code?

**Comment by rebeccahum** [comment]
Sure, the one illustrated above throws a false positive.  That should not be flagged.

**Comment by GaryJones** [comment]
I updated the OP to fix non-related violations in the demo code.

Running it with:

```sh
.composer/vendor/bin/phpcs --standard=WordPress-Extra,WordPress-VIP-Go ~/test.php -s
```

Gave me:

> 11 | ERROR | Use of `static::$_type` inside closure. (WordPressVIPMinimum.Variables.VariableAnalysis.StaticOutsideClass)

Which seems correct - there is a `static::` in the closure.

I don't see any violation related the `self::`.

**Comment by david-binda** [comment]
I believe that this issue should be reported [upstream](https://github.com/sirbrillig/phpcs-variable-analysis) and handled there.

The `static`, `self` and `$this` inside closures in PHP prior to 5.4.0 produced fatal errors. But PHP 5.4.0 automatically bounds current class' scope to closure, making `$this`, `self`, and `static` available inside of the function's scope.

See https://3v4l.org/B43Uj and https://3v4l.org/sdJ6Q with code examples and their processing in different PHP versions (works as expected in PHP 5.4+).

The `self` part of the sniff was introduced [8 years ago](https://github.com/sirbrillig/phpcs-variable-analysis/commit/372456c599c34b9096b85e6efa6a088e8e38b760) ( 20 Jul 2011
 ) prior the release of PHP 5.4.0 (which happened on 01 Mar 2012 ).

But really, this is more for upstream, which I'll do shortly.

Let me copy the code examples here, for more posterity:

```
<?php

class My_Class {

    protected static $output = 'test';

    public static function method() {
        return function() {
            echo static::$output;
        };
    }
}

$object = new My_Class;
$function = $object->method();
$function();
```

and

```
<?php

class My_Class {

    protected static $output = 'test';

    public static function method() {
        return function() {
            echo self::$output;
        };
    }
}

$object = new My_Class;
$function = $object->method();
$function();
```

**Comment by david-binda** [comment]
https://github.com/sirbrillig/phpcs-variable-analysis/issues/68

**Comment by GaryJones** [comment]
> I believe that this issue should be reported [upstream](https://github.com/sirbrillig/phpcs-variable-analysis) and handled there.

Please note: https://github.com/Automattic/VIP-Coding-Standards/issues/236 :-)

**Comment by rebeccahum** [comment]
Looks like it's been fixed upstream now but our version is out of sync with upstream.  Perhaps it's time for syncing with upstream, @GaryJones?

**Comment by GaryJones** [comment]
Thank you for demos - I now understand that the code works fine, and the violation no longer needs to be reported (for PHP 5.4 and later at least).

I think there are more benefits to be gained by pulling in @sirbrillig's package, so let's keep this open until that's done, but no separate action should be needed here. The other ticket can focus on where we might already have made changes to our current version.[](url)

**Comment by GaryJones** [comment]
Now that #449 is completed to use VariableAnalysis proper, I've confirmed that the provided examples above don't give an error with the `develop` version of `WordPress-VIP-Go`, so this is now fixed for VIPCS 2.2.0.
